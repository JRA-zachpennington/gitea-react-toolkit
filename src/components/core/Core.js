import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { Paper, Button } from '@material-ui/core';

import { useAuthentication } from '..';

function Core({
  promise,
  props,
  authenticate,
}) {
  const [authentication, setAuthentication] = useState(null);
  const [response, setResponse] = useState();
  const [loading, setLoading] = useState(false);
  const { component: authenticationComponent } = useAuthentication({ config: props.config, onAuthentication: setAuthentication });
  const needsAuthentication = (authenticate && !authentication && authenticationComponent);
  const _promise = useCallback(async () => {
    try {
      setLoading(true);
      const _response = await promise({
        ...props,
        ...authentication,
      });
      setResponse(_response);
    } catch (error) {
      const _error = { error: error.message };
      setResponse(_error);
    }
    setLoading(false);
  }, [authentication, promise, props]);

  let responseComponent = <div />;

  if (needsAuthentication && authenticationComponent) {
    return authenticationComponent;
  }

  if (response) {
    responseComponent = <div>{JSON.stringify(response)}</div>;
  } else if (loading) {
    responseComponent = <div/>;
  } else if (!response && typeof (response) !== 'undefined' && !loading) {
    responseComponent = 'No response';
  }
  return (
    <>
      <h3>Props</h3>
      <Paper>
        <div>{JSON.stringify(props)}</div>
      </Paper>
      <h3>Response</h3>
      <Paper>
        {responseComponent}
      </Paper>
      <Button variant="contained" color='primary' onClick={_promise}>
        Run
      </Button>
    </>
  );
};

Core.propTypes = {
  props: PropTypes.object.isRequired,
  promise: PropTypes.func.isRequired,
  authenticate: PropTypes.bool,
};

Core.defaultProps = {
  authenticate: false,
  confirm: false,
};

export default Core;