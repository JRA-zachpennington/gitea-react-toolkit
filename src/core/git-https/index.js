export * from './gitFile';
export * from './gitApi';
export * from './gitZip';
