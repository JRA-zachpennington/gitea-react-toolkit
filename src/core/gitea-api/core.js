import localforage from 'localforage';
import axios from 'axios';
import { setup } from 'axios-cache-adapter';

export const apiPath = 'api/v1';

export const get = async ({url, params, config={}, noCache}) => {
  const _config = {baseURL: config.server, ...config};
  let response;
  if (noCache) {
    const _params = {noCache: Math.random(), ...params};
    response = await axios.get(url, {..._config, params: _params});
  }
  else response = await api.get(url, {..._config, params});
  const {data} = response;
  return data;
};

export const post = async ({url, payload, config}) => {
  const _config = {baseURL: config.server, ...config};
  const {data} = await axios.post(url, payload, _config);
  return data;
};

export const put = async ({url, payload, config}) => {
  const _config = {baseURL: config.server, ...config};
  const {data} = await axios.put(url, payload, _config);
  return data;
};

export const patch = async ({url, payload, config}) => {
  const _config = {baseURL: config.server, ...config};
  const {data} = await axios.patch(url, payload, _config);
  return data;
};

export const del = async ({url, payload, config}) => {
  const _config = {baseURL: config.server, ...config};
  config.data = payload;
  const {data} = await axios.delete(url, _config);
  return data;
};

const cacheStore = localforage.createInstance({
  driver: [localforage.INDEXEDDB],
  name: 'web-cache',
});

const api = setup({
  crossDomain: true,
  cache: {
    store: cacheStore,
    maxAge: 1 * 1 * 1 * 1 * 1000,
    exclude: { query: false },
    key: req => {
      // if (req.params) debugger
      let serialized = req.params instanceof URLSearchParams ?
      req.params.toString() : JSON.stringify(req.params) || '';
      return req.url + serialized;
    },
  },
});