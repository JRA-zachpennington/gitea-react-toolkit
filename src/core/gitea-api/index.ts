export * from './core';
export * from './authentication';
export * from './users';
export * from './repos';
