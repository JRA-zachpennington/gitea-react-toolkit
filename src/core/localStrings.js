//
// Translate strings into different languages
// Then use localString(id) to return the language specific string to use
//
const strings = {
    en:{
        OpenRepo: "Go to Repo",
        OpenOrg: "Go to Organization",
        StringMissing: "String Id missing",
        LangNotSupported:  "Language not supported"
    },
};

export const localString = (id) => {
    let lang = null;
    
    if (typeof window !== 'undefined' )
    {
        if (navigator) {
            lang = navigator.language.split(/-|_/)[0];
        }
    }
    // if language is unknown (not sure this can actually happen)
    if (!lang) {
        lang = 'en';
    }
    // if there are no strings for the language
    if ( strings[lang] === undefined ) {
        lang = 'en';
        return strings[lang]['LangNotSupported'];
    }
    let lstring = strings[lang][id];
    // if a string for the message id is missing
    if ( lstring === undefined ) {
        return strings[lang]['StringMissing']
    }
    return lstring;
};
