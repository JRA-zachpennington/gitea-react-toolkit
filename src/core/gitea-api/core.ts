import localforage from 'localforage';
import axios from 'axios';
import { setup } from 'axios-cache-adapter';
import { authorizationHeaders } from './authentication';
import { APIConfig } from './core.d';
import { AuthToken } from './index.d';
export const apiPath = 'api/v1';
const DEFAULT_MAX_AGE = 1000;

const cacheStore = localforage.createInstance({
  driver: [localforage.INDEXEDDB],
  name: 'web-cache',
});

const api = setup({
  crossDomain: true,
  cache: {
    store: cacheStore,
    maxAge: 1 * 1 * 1 * 1 * DEFAULT_MAX_AGE,
    exclude: { query: false },
    key: req => {
      const serialized = req.params instanceof URLSearchParams ?
        req.params.toString() : JSON.stringify(req.params) || '';
      return req.url + serialized;
    },
  },
});

interface RestAPI {
  config: ExtendConfig;
  url: string;
  payload: object;
}

export interface ExtendConfig {
  token: AuthToken;
  tokenid: string;
  headers?: object;
  server?: string;
  data?: object;
}

export const extendConfig = (config: ExtendConfig): APIConfig => {
  const authHeaders = authorizationHeaders({ token: config.token });
  const headers = { ...config.headers, ...authHeaders };
  const _config = {
    baseURL: config.server, ...config, headers,
  };
  return _config;
};

interface Get {
  config: ExtendConfig;
  url: string;
  params?: object;
  noCache?: number | boolean;
}

export const get = async ({
  url, params = {}, config, noCache,
}: Get): Promise<any> => {
  const _config = extendConfig(config);
  let response;

  if (noCache) {
    const _params = { noCache: Math.random(), ...params };
    response = await axios.get(url, { ..._config, params: _params });
  } else {
    response = await api.get(url, { ..._config, params });
  }

  const { data } = response;
  return data;
};

export const post = async ({
  url, payload, config,
}: RestAPI): Promise<object> => {
  const _config = extendConfig(config);
  const { data } = await axios.post(url, payload, _config);
  return data;
};

export const put = async ({
  url, payload, config,
}: RestAPI): Promise<object> => {
  const _config = extendConfig(config);
  const { data } = await axios.put(url, payload, _config);
  return data;
};

export const patch = async ({
  url, payload, config,
}: RestAPI): Promise<object> => {
  const _config = extendConfig(config);
  const { data } = await axios.patch(url, payload, _config);
  return data;
};

export const del = async ({
  url, payload, config,
}: RestAPI): Promise<object> => {
  const _config = extendConfig(config);
  config.data = payload;
  const { data } = await axios.delete(url, _config);
  return data;
};