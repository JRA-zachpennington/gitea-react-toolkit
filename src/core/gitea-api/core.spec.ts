/* eslint-disable @typescript-eslint/no-magic-numbers */
/// <reference types="jest" />
import * as helpers from './core';
jest.mock('axios-cache-adapter', () => ({
  setup: () => ({
    get: (url) =>
      url.match(/fail/ig) ? Promise.reject(new Error('Request failed with status code 404')) : Promise.resolve({ data: 'OK' }),
  }),
}));

const TEST_TOKEN = 'encrypted123456789';
const authToken = {
  sha1: TEST_TOKEN,
  id: 'test-id',
  name: 'test-token',
};
const config = {
  token: authToken,
  tokenid: TEST_TOKEN,
  server: 'test-server',
};

describe('extendConfig', () => {
  it('should pass', () => {
    const res = helpers.extendConfig(config);

    expect(res).toEqual(expect.objectContaining({
      baseURL: config.server,
      headers: expect.objectContaining({
        'Authorization': expect.stringMatching(/token\s/),
        'Content-Type': 'application/json',
      }),
      token: {
        sha1: TEST_TOKEN, id: authToken.id, name: authToken.name,
      },
      tokenid: config.tokenid,
    }));
  });
});

describe('get', () => {
  it('should fail with 404', () => {
    const params = {
      config,
      url: 'https://failing.com',
    };

    return expect(helpers.get(params)).rejects.toMatchSnapshot();
  });

  it('should pass', () => {
    const params = {
      config,
      url: 'https://passing.com',
    };

    return expect(helpers.get(params)).resolves.toEqual('OK');
  });
});
