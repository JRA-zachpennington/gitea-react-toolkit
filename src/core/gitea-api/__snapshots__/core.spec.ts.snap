// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get should fail with 404 1`] = `[Error: Request failed with status code 404]`;
